### PR TITLE
docs: Core concepts page

### PR DESCRIPTION
**Pull request by bpmutter**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Add a page to the user guide about the core concepts of ESLint.

The page contains minimalistic explanations, with links out to the more detailed explanation pages. 

#### Is there anything you'd like reviewers to focus on?

I intentionally kept this page pretty minimalist. Specifically, I didn't include any code examples. I decided not to include code examples to really keep this page focused on the *concepts* (pure [explanation using the Divio framework](https://documentation.divio.com/explanation/)). I'm not 100% convinced this is the correct decision, and look forward to the reviewers' thoughts and feedback on the matter.

Fixes https://github.com/eslint/eslint/issues/16382

### CODE REVIEW

File: docs/src/user-guide/core-concepts.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    title: Core Concepts
+    parent: user guide
+    order: 2
+---
+
+This page contains a high-level overview of some of the core concepts of ESLint.
+
+## What is ESLint?
+
+ESLint is a configurable JavaScript linter. It's like spell check and autocorrect
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:15]: I'd like to put less emphasis on consistent styling and focus on ESLint as a general purpose tool to identify certain patterns in your code. 

Maybe something like "It helps you find and fix problems in your JavaScript code. Problems can be anything from potential runtime bugs, to not following best practices,  to styling issues."?

File: docs/src/user-guide/core-concepts.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+for your JavaScript code. It helps you write JavaScript in a consistent style
+and fix bugs in your code.
+
+## Rules
+
+Rules are the core building block of ESLint. A rule validates if your code meets a
+certain expectation, and what to do if it does not meet that expectation.
+Rules can also contain additional configuration options specific to that rule.
+
+For example, the [`semi` rule](../rules/semi.md`) lets you specify whether or not
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:25]: We generally only link the rule name in the docs. Also, `.md` links won't work when rendered into the website, so you can just drop the extension:

```suggestion
For example, the [`semi`](../rules/semi`) rule lets you specify whether or not
```

File: docs/src/user-guide/core-concepts.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+For more information, refer to [Rules](../rules/).
+
+## Configuration Files
+
+An ESLint configuration file is a single place where you put all the configuration
+for ESLint your project. You can include built-in rules, how you want them enforced,
+plugins with custom rules, which files you want rules to apply to, and more.
+
+For more information, refer to [Configuration Files](./configuring/configuration-files.md).
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:42]:
```suggestion
For more information, refer to [Configuration Files](./configuring/configuration-files).
```

File: docs/src/user-guide/core-concepts.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+ESLint contains hundreds of built-in rules that you can use.
+You can also create custom rules or use rules that others have
+created with [plugins](#plugins).
+
+For more information, refer to [Rules](../rules/).
+
+## Configuration Files
+
+An ESLint configuration file is a single place where you put all the configuration
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:38]: Technically not a single place. You can actually have multiple throughout your project and ESLint will merge them all for you. We are moving away from this with the new config system, which does have only one file that it uses.

File: docs/src/user-guide/core-concepts.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+for ESLint your project. You can include built-in rules, how you want them enforced,
+plugins with custom rules, which files you want rules to apply to, and more.
+
+For more information, refer to [Configuration Files](./configuring/configuration-files.md).
+
+## Plugins
+
+An ESLint plugin is an npm module that contains a set of ESLint rules.
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:46]: Plugins can contain rules, configurations, processors, and environments.

File: docs/src/user-guide/core-concepts.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+Plugins can be used to enforce a style guide and support JavaScript extensions
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:47]: Typically shareable configs are used to enforce style guides using built-in rules while plugins provide additional rules.

A popular common use case is to enforce best practices for a given framework (React, Vue, and Angular all use ESLint plugins to provide rules specific to how best to use those frameworks).

File: docs/src/user-guide/core-concepts.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+You can add single rules or sets of rules from a plugin.
+Often plugins come with recommended sets of rules that you can include in your project.
+
+For more information, refer to [Configuring Plugins](./configuring/plugins.md).
+
+## Formatters
+
+An ESLint formatter controls the appearance of the linting results.
+If you're using a code editor with an ESLint integration to see linting results
+in line, the editor is using an ESLint formatter to display those results.
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:58]: Inline messages, such as those in the VS Code editor, don't use formatters.

Formatters are solely for use with the command line interface.

File: docs/src/user-guide/core-concepts.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+For more information, refer to [Formatters](./formatters/index.md).
+
+## Integrations
+
+One of the things that makes ESLint such a useful tool is the ecosystem of integrations
+that surrounds. These integrations extend the core functionality of ESLint.
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:65]: I don't think integrations extend the core, but rather integrations build ESLint right into your editor, build tool, IDE, etc. So they are extending the host of ESLint, not ESLint itself.

File: docs/src/user-guide/core-concepts.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+For example, many code editors have ESLint extensions that show you the ESLint results
+of your code in the file as you work so that you don't need to use the ESLint CLI
+to see linting results.
+
+For more information, refer to [Integrations](./integrations.md).
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:70]:
```suggestion
For more information, refer to [Integrations](./integrations).
```

File: docs/src/user-guide/core-concepts.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+For a curated collection of popular ESLint integrations,
+see the [awesome-eslint Github repository](https://github.com/dustinspecker/awesome-eslint).
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:72]: It seems like this is primarily about plugins and configs and doesn't have any obvious section for integrations. Did you mean to include this in the plugins section?

File: docs/src/user-guide/core-concepts.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+The ESLint CLI is a command line interface that lets you execute linting
+from the terminal. The CLI has a variety of options that you can pass to its commands.
+
+The ESLint Node.js API lets you use ESLint programmatically from Node.js code.
+The API is useful when developing plugins, integrations, and other tools related
+to ESLint.
+
+Unless you are extending ESLint in some way, you should use the CLI.
+
+For more information, refer to [Command Line Interface](./command-line-interface.md)
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:85]:
```suggestion
For more information, refer to [Command Line Interface](./command-line-interface)
```

File: docs/src/user-guide/core-concepts.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+and [Node.js API](../developer-guide/nodejs-api.md).
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:86]:
```suggestion
and [Node.js API](../developer-guide/nodejs-api).
```

**Review by nzakas — CHANGES_REQUESTED** [review]
Thanks for putting this together! I left some small notes throughout.

I think there are probably some other core concepts that would be worth adding to this page:

1. Shareable configs - this is the largest part of the ESLint ecosystem, so I think it makes sense to mention this after configuration files.
2. Parsers - many folks use alternate parsers to help ESLint parse different syntax. `typescript-eslint` (for TypeScript syntax) and `@babel/eslint-parser` (for experimental syntax) are examples.
3. Processors - allow ESLint to parse JS out of another file. For example `eslint-plugin-markdown` allows ESLint to pull JS code out of Markdown files and lint it.

**Comment by nzakas** [comment]
I just noticed - it appears that your editor may be adding in hard line breaks after a certain number of characters on each line. Can you adjust that so those hard line breaks don't appear? At least in my editor, this isn't the default, so it's a bit harder for me to make changes because the line breaks move around as the content changes.

File: docs/src/user-guide/core-concepts.md (hunk 1 above)
Comment (bpmutter) [reply docs/src/user-guide/core-concepts.md:15]: > It helps you find and fix problems in your JavaScript code. Problems can be anything from potential runtime bugs, 
> to not following best practices, to styling issues.

love this. using

File: docs/src/user-guide/core-concepts.md (hunk 6 above)
Comment (bpmutter) [reply docs/src/user-guide/core-concepts.md:47]: thx, will update accordingly

File: docs/src/user-guide/core-concepts.md (hunk 8 above)
Comment (bpmutter) [reply docs/src/user-guide/core-concepts.md:65]: will rephrase

File: docs/src/user-guide/core-concepts.md (hunk 10 above)
Comment (bpmutter) [reply docs/src/user-guide/core-concepts.md:72]: i included this b/c it was [on the Integration page](https://eslint.org/docs/latest/user-guide/integrations#other-integration-lists), and seemed like a good collection of resources. 

we can just remove the link here (and then laster look at if it should be on the Integrations page.

File: docs/src/user-guide/core-concepts.md (hunk 7 above)
Comment (bpmutter) [reply docs/src/user-guide/core-concepts.md:58]: ok, thanks for clarifying. what are some reasons you might want to use a formatter? 

for example, why would i want to output my results to [HTML](https://eslint.org/docs/latest/user-guide/formatters/#html)?

**Comment by bpmutter** [comment]
> I just noticed - it appears that your editor may be adding in hard line breaks after a certain number of characters on each line. Can you adjust that so those hard line breaks don't appear? At least in my editor, this isn't the default, so it's a bit harder for me to make changes because the line breaks move around as the content changes.

as a personal practice i break lines (by hitting enter manually) around 80 characters. this lets you have a more specific commit history and looks better if the editor doesn't have word wrapping. 

but will refactor to stick by this projects conventions and make `1 paragraph == 1 line`

**Comment by bpmutter** [comment]
@nzakas implemented all your feedback. please take another look when you get the chance!

File: docs/.vscode/settings.json [hunk 13]
```diff
@@ -0,0 +1,3 @@
+{
+  "editor.formatOnType": false
+}
```
Comment (nzakas) [inline docs/.vscode/settings.json:3]: We aren’t typically including vscode settings file. Was this included by mistake?

File: docs/src/user-guide/core-concepts.md [hunk 14]
```diff
@@ -12,75 +12,72 @@ This page contains a high-level overview of some of the core concepts of ESLint.
...
-Often plugins come with recommended sets of rules that you can include in your project.
+An ESLint plugin is an npm module that can contain a set of ESLint rules, configurations, processors, and environments. Often plugins include custom rules. Plugins can be used to enforce a style guide and support JavaScript extensions (like TypeScript), libraries (like React), and frameworks (Angular).
+
+A popular use case for plugins is to enforce best practices for a framework. For example, [@angular-eslint/eslint-plugin](https://www.npmjs.com/package/@angular-eslint/eslint-plugin) contains best practices for using the Angular framework.
 
 For more information, refer to [Configuring Plugins](./configuring/plugins.md).
 
+## Parsers
+
+An ESLint parser converts code into an abstract syntax tree that ESLint can evaluate. By default, ESLint uses a built-in parser that is compatible with standard JavaScript runtimes and versions.
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:51]: Might be worth mentioning the default parser is Espree and link to the eslint/espree repo in this paragraph.

File: docs/src/user-guide/core-concepts.md [hunk 15]
```diff
@@ -12,75 +12,72 @@ This page contains a high-level overview of some of the core concepts of ESLint.
...
+
+Custom parsers let ESLint parse non-standard JavaScript syntax. Often custom parsers are included as part of shareable configurations or plugins, so you don't have to use them directly.
+
+For example, [@typescript-eslint/parser](npmjs.com/package/@typescript-eslint/parser) is a custom parser included in the [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) project that lets ESLint parse TypeScript code.
+
+## Custom Processors
+
+An ESLint Processor extracts JavaScript code from other kinds of files, then lets ESLint lint the JavaScript code. Alternatively, you can use a processor to manipulate JavaScript code before parsing it with ESLint.
```
Comment (nzakas) [inline docs/src/user-guide/core-concepts.md:59]:
```suggestion
An ESLint processor extracts JavaScript code from other kinds of files, then lets ESLint lint the JavaScript code. Alternatively, you can use a processor to manipulate JavaScript code before parsing it with ESLint.
```

**Review by nzakas — CHANGES_REQUESTED** [review]
Nice, this is looking really good. I just left a couple of small comments and then we should be good to go.

File: docs/.vscode/settings.json (hunk 13 above)
Comment (bpmutter) [reply docs/.vscode/settings.json:3]: yep, will remove!

**Comment by bpmutter** [comment]
@nzakas implemented your feedback. ready for re-review

**Review by nzakas — APPROVED** [review]
Awesome! Very excited to get this out. Thanks so much.

**Pull request merged**